katsdptelstate/rdb_writer.py

```diff
@@ -16,145 +16,146 @@
-    """
-    def __init__(self, endpoint=None, client=None):
-        if not endpoint and not client:
-            raise ValueError("You must specify either an endpoint or a valid client.")
-        if client:
-            self._r = client
-        else:
-            import redis
-            if isinstance(endpoint, str):
-                endpoint = endpoint_parser(DEFAULT_PORT)(endpoint)
-            self._r = redis.StrictRedis(host=endpoint.host, port=endpoint.port)
-        self.logger = logging.getLogger(__name__)
-
-    def save(self, filename, keys=None, supplemental_dumps=[]):
-        """Encodes specified keys from the RDB file into binary
-        string representation and writes these to a file.
+    Subsequent bytes represent the name of the key, which consists of a length
+    encoding followed by the actual byte representation of the string. For this simple
+    writer we only provide two length encoding schemas:
 
-        Parameters
-        ----------
-        filename : str
-            Destination filename. Will be opened in 'wb'.
-        keys : list
-            A list of the keys to extract from Redis and include in the dump.
-            Keys that don't exist will not raise an Exception, only a log message.
-            None (default) includes all keys.
-        supplemental_dumps : list
-            A list of encoded supplemental key/values in string form to be included
-            in the RDB dump.
-
-        Returns
-        -------
-        keys_written : int
-            Number of keys written to the file.
-        keys_failed : int
-            Number of keys that failed to be written.
-        """
-        if keys is None:
-            self.logger.warning("No keys specified - dumping entire database")
-            keys = self._r.keys(b'*')
-
-        with open(filename, 'wb') as f:
-            f.write(RDB_HEADER)
-            keys_written = 0
-            keys_failed = 0
-            for key in keys:
-                try:
-                    enc_str = self.encode_item(key)
-                except (ValueError, KeyError) as e:
-                    keys_failed += 1
-                    self.logger.error("Failed to save key: {}".format(e))
-                    continue
-                f.write(enc_str)
-                keys_written += 1
-            for dump in supplemental_dumps:
-                f.write(dump)
+    - String length up to 63 - single byte, 6 LSB encode number directly
+    - String length from 64 to 16383 - two bytes, 6 LSB of byte 1 + 8 from byte 2 encode number
+
+    Thereafter the value, encoding according to its appropriate schema is appended.
+    As a shortcut, the Redis DUMP command is used to generate the encoded value string.
+
+    Note: Redis provides a mechanism for optional key expiry, which we ignore here.
+    """
+    try:
+        key_len = encode_len(len(key))
+    except ValueError as e:
+        raise ValueError('Failed to encode key length: {}'.format(e))
```

Comment:
On Python 3 this should probably be `raise ... from None` (wrapped up `six.raise_from` of course).

Reply:
Great, done. I wondered if there was a way to suppress the context :-)